### ISSUE TITLE
Add select operator

### ISSUE THREAD

**Issue by jpsamaroo** [JuliaParallel/Dagger.jl#309]
We should implement a `select` operator, which returns the result (and probably ID) of the first task to complete out of a set of tasks. This operator should have the option to cancel unfinished tasks, and to either `fetch` or `wait`.

**Comment by vchuravy** [comment]
cc: @tkf

**Comment by tkf** [comment]
@vchuravy told me there are UCX atomic ops http://openucx.github.io/ucx/api/v1.5/html/group___u_c_t___a_m_o.html that I could use for this.

So, if you don't mind a deep stack of yak shaving, an interesting approach would be to implement [Reagents.jl](https://github.com/tkf/Reagents.jl) on top of the CAS API from UCX.jl. But the CAS'ed values are typically object references. I don't know what the object reference would be in the multi-node setting. Maybe it'd be something like the chunk id or something?

**Comment by vchuravy** [comment]
I don't think we need to recurse into that just yet. @jpsamaroo are these tasks that are both local and remote?

**Comment by tkf** [comment]
Yeah, directly implementing this seems to be a good direction as a first cut. Also, re-reading the OP, I think it's not really the CSP style selective communication (e.g., as in Go) where it executes one and *only* one synchronization. I think the API in the OP is similar to Python's [`asyncio.wait(awaitables, return_when=FIRST_COMPLETED)`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait). So maybe something like `waitfirst(thunks)` is a better name.

**Comment by jpsamaroo** [comment]
Yeah, it's not only-one semantics, but first-to-finish semantics. `waitfirst` sounds good!

> are these tasks that are both local and remote?

They can be executing anywhere, but we have a valid handle (`EagerThunk`, `Thunk`, or `ThunkID`) to them that's passed to `waitfirst`.
